doc/known-warnings.txt

```diff
@@ -16,3 +16,6 @@
 .*Duplicate C declaration.*\n.*'\.\. c:.*:: .*struct in6_addr.*'.*
 .*Duplicate C declaration.*\n.*'\.\. c:.*:: .*struct net_if.*'.*
 .*Duplicate C declaration.*\n.*'\.\. c:struct:: bt_ots_init'.*
+.*Duplicate C declaration.*\n.*'\.\. c:member:: enum *bt_mesh_dfd_upload_phase bt_mesh_dfd_srv.phase'.*
+.*Duplicate C declaration.*\n.*'\.\. c:member:: struct *bt_mesh_blob_xfer bt_mesh_dfu_cli.blob'.*
+.*Duplicate C declaration.*\n.*'\.\. c:member:: struct *net_if *\* net_if_mcast_monitor.iface'.
```

Comment:
Initial impression is that the root of the issue at times seems to be that the Sphinx C-domain does not have good handling for when structs, functions and other entities have the same name. This is allowed in C and present a few times in the Zephyr code base. This leads to duplicate targets being generated and triggering warnings from the Sphinx C Domain  code.

It is mentioned here: https://github.com/sphinx-doc/sphinx/issues/8241#issuecomment-701245203 in an issue which goes on to reference this pull request: https://github.com/sphinx-doc/sphinx/pull/8313 as the solution but that seems to have been open for a long time. 

As the references are parsed and stored by the C Domain which then generates the warnings for clashes, I don't see there being a clear way to work around them other than the current regexes. The question then becomes the difference between Breathe and Docleaf and that comes down to how much of the Doxygen XML the two systems attempt to render. In various edge cases, the XML becomes harder to make clear sense of and so the two systems potentially draw the lines at different points in different places. Neither are perfect. 

I will work through the various cases (as highlighted by deleting all the lines in known-warnings.txt file) and document why each appears to be happening and attempt to improve Docleaf where possible. Any due to name clashes are potentially going to have to stay but I can see at least one case where field name is clashing with a field in a nested structure and that might be possible to avoid.

Reply:
I've added a new commit which adds categories to the known-warnings.txt file attempting to indicate the source of the various issues. Some are due to entities sharing the same name and the Sphinx issues indicated before. Some are due to fields with the same name appearing in multiple anonymous structs. Some are due to Doxygen not indicating `extern` declarations in the XML output and so Docleaf and Breathe are unable to filter them out in any fashion. I have raised a ticket for this: https://github.com/doxygen/doxygen/issues/10163

It is possible to use the Doxygen `@cond` syntax to have it ignore parts of the file and so we could use that to explicit stop it processing the `extern` declarations but that would introduce more changes than the current known-warnings.txt patterns.

Finally the new lines are due to Docleaf attempting to render fields from anonymous structs which then clash with fields in the parent struct. I believe this didn't happen in Breathe as it doesn't attempt to render the members in some of these situations. For example, the breathe output for `bt_mesh_dfd_srv` is:

![image](https://github.com/zephyrproject-rtos/zephyr/assets/5390/126e6c61-9809-49b5-908a-ba9accc7ff76)

Whilst the docleaf output is:

![image](https://github.com/zephyrproject-rtos/zephyr/assets/5390/005e2d6e-ba31-4552-9c42-ff87a05e11f8)

With the actual code being: https://github.com/zephyrproject-rtos/zephyr/blob/db1a718341a3724b82c052e6e6b7db19251f7a22/include/zephyr/bluetooth/mesh/dfd_srv.h#L117-L136

Which shows that the anonymous struct members incorrectly appear in the body of the main struct. Unfortunately this is because the Doxygen xml output ([structbt__mesh__dfd__srv.zip](https://github.com/zephyrproject-rtos/zephyr/files/11930141/structbt__mesh__dfd__srv.zip)) doesn't distinguish between the two levels. It flattens everything. I'm not sure what is better or worse. Leaving out the details or having them poorly represented. I haven't yet looked to see if there is a doxygen issue for this but I fear it might involve changes to their XML structures to better support this and that seems like it is possibly unlikely.

Reply:
I would welcome input as to what we might want to address or account for at this stage or if the current state is acceptable in some fashion for moving forwards.